src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -488,25 +578,45 @@ private void ExtractAsRegularFile(string destinationFileName)
                     // Important: The DataStream will be written from its current position
                     DataStream.CopyTo(fs);
                 }
-                SetModeOnFile(fs.SafeFileHandle, destinationFileName);
+                SetModeOnFile(fs.SafeFileHandle);
             }
 
             ArchivingUtils.AttemptSetLastWriteTime(destinationFileName, ModificationTime);
         }
 
-        // Abstract method that extracts the current entry when it is a block device.
-        partial void ExtractAsBlockDevice(string destinationFileName);
-
-        // Abstract method that extracts the current entry when it is a character device.
-        partial void ExtractAsCharacterDevice(string destinationFileName);
-
-        // Abstract method that extracts the current entry when it is a fifo file.
-        partial void ExtractAsFifo(string destinationFileName);
+        // Asynchronously extracts the current entry as a regular file into the specified destination.
+        // The assumption is that at this point there is no preexisting file or directory in that destination.
+        private async Task ExtractAsRegularFileAsync(string destinationFileName, CancellationToken cancellationToken)
+        {
+            Debug.Assert(!Path.Exists(destinationFileName));
 
-        // Abstract method that extracts the current entry when it is a hard link.
-        partial void ExtractAsHardLink(string targetFilePath, string hardLinkFilePath);
+            FileStreamOptions fileStreamOptions = new()
+            {
+                Access = FileAccess.Write,
+                Mode = FileMode.CreateNew,
+                Share = FileShare.None,
+                PreallocationSize = Length,
+                Options = FileOptions.Asynchronous
+            };
+            // Rely on FileStream's ctor for further checking destinationFileName parameter
+            using (FileStream fs = new FileStream(destinationFileName, fileStreamOptions))
+            {
+                if (DataStream != null)
+                {
+                    // Important: The DataStream will be written from its current position
+                    await DataStream.CopyToAsync(fs, cancellationToken).ConfigureAwait(false);
+                }
+                try
+                {
+                    SetModeOnFile(fs.SafeFileHandle);
+                }
+                catch (Exception e)
+                {
+                    await Task.FromException(e).ConfigureAwait(false);
```

Comment:
This is wrong, isn't it?

I'm just reading the comment by @bartonjs here: https://github.com/dotnet/runtime/issues/70905#issuecomment-1159117124

So I assume I need to catch exceptions in the calling method, not here?

Reply:
I don't understand what this is trying to do.  As written it's an expensive nop that eats all exceptions thrown.

This is all inside of an async method.  An async method catches all exceptions that go unhandled in the body and stores them into the Task being returned.  Which is what you want here... just remove the try/catch and let any exceptions propagate.

I think you're making it harder than it actually is.  In general, let exceptions propagate exactly as you normally would in a synchronous method.  The only special-cases typically are when doing argument validation in the entry point method.

Reply:
> I think you're making it harder than it actually is

I know, that's why I'm asking 🙂. There are many things about async code that I am unfamiliar with, so I appreciate the patience and the explanations.

> As written it's an expensive nop that eats all exceptions thrown.

Ah alright, so I had the wrong understanding.

> In general, let exceptions propagate exactly as you normally would in a synchronous method. The only special-cases typically are when doing argument validation in the entry point method.

Thanks, that is the answer I was looking for. I'll adjust the code based on this.